src/ng/parse.js

```diff
@@ -622,15 +622,51 @@ function isStateless($filter, filterName) {
   return !fn.$stateful;
 }
 
-function findConstantAndWatchExpressions(ast, $filter) {
+// Detect nodes which could depend on non-shallow state of objects
+function isShallowState(stack) {
+  for (var i = 0; i < stack.length; i++) {
+    var node = stack[i];
+    switch (node.type) {
+      // Computed members might invoke a stateful toString()
+      case AST.MemberExpression:
+        if (node.computed) {
+          return false;
+        }
+        break;
+
+      // The + operator can invoke a stateful toString().
+      // Otherwise operators can not access any state within objects.
+      case AST.UnaryExpression:
+      case AST.BinaryExpression:
+        if (node.operator === '+') {
+          return false;
+        }
+        break;
+
+      // Functions / filters probably read state from within objects
+      case AST.CallExpression:
+        return false;
+    }
+  }
+
+  return true;
+}
+
+function findConstantAndWatchExpressions(ast, $filter, stack) {
```

Comment:
I am not following completely the logic here, but by just looking at the code it looks like `stack` can be replace with a simple boolean (that the first called call with `true` and the rest do `ast.isShallowState = parentShallowState && isShallowState([ast]);` (and of course this can be simplified).